osu.Game/Skinning/LegacySkin.cs

```diff
@@ -433,32 +437,44 @@ public override SampleChannel GetSample(ISampleInfo sampleInfo)
             return null;
         }
 
-        private IEnumerable<string> getFallbackNames(string componentName)
-        {
-            // May be something like "Gameplay/osu/approachcircle" from lazer, or "Arrows/note1" from a user skin.
-            yield return componentName;
-
-            // Fall back to using the last piece for components coming from lazer (e.g. "Gameplay/osu/approachcircle" -> "approachcircle").
-            string lastPiece = componentName.Split('/').Last();
-            yield return componentName.StartsWith("Gameplay/taiko/", StringComparison.Ordinal) ? "taiko-" + lastPiece : lastPiece;
-        }
-
         private IEnumerable<string> getLegacyLookupNames(HitSampleInfo hitSample)
         {
-            var lookupNames = hitSample.LookupNames;
+            var lookupNames = hitSample.LookupNames.SelectMany(getFallbackNames);
 
             if (!UseCustomSampleBanks && !string.IsNullOrEmpty(hitSample.Suffix))
+            {
                 // for compatibility with stable, exclude the lookup names with the custom sample bank suffix, if they are not valid for use in this skin.
                 // using .EndsWith() is intentional as it ensures parity in all edge cases
                 // (see LegacyTaikoSampleInfo for an example of one - prioritising the taiko prefix should still apply, but the sample bank should not).
-                lookupNames = hitSample.LookupNames.Where(name => !name.EndsWith(hitSample.Suffix, StringComparison.Ordinal));
+                foreach (var l in lookupNames)
+                {
+                    if (!l.EndsWith(hitSample.Suffix, StringComparison.Ordinal))
+                    {
+                        foreach (var n in getFallbackNames(l))
```

Comment:
Is this necessary? Isn't it effectively `getFallbackNames(getFallbackNames(lookup))`?

I'm not fully understanding this code, could it be written as:

```
var lookupNames = hitSample.LookupNames.SelectMany(getFallbackNames);

if (!UseCustomSampleBanks && !string.IsNullOrEmpty(hitSample.Suffix))
{
    // for compatibility with stable, exclude the lookup names with the custom sample bank suffix, if they are not valid for use in this skin.
    // using .EndsWith() is intentional as it ensures parity in all edge cases
    // (see LegacyTaikoSampleInfo for an example of one - prioritising the taiko prefix should still apply, but the sample bank should not).
    lookupNames = hitSample.LookupNames.Where(name => !name.EndsWith(hitSample.Suffix, StringComparison.Ordinal));
}

foreach (var l in lookupNames)
    yield return l;

// also for compatibility, try falling back to non-bank samples (so-called "universal" samples) as the last resort.
// going forward specifying banks shall always be required, even for elements that wouldn't require it on stable,
// which is why this is done locally here.
yield return hitSample.Name
```

Reply:
Good catch – I moved the `getFallbackNames` out and forgot to remove this.

I'm fine without the way you have it, or just removing the nested call (not sure if yield is more efficient than `.Where`, that's the only reason I was using it).

Reply:
it's probably about the same either way.